File: sort/quick/quicksort.js
"use strict";

function quickSort(arr, start, end) {
    if (start < end) {
        let pIndex = partition(arr, start, end);
        quickSort(arr, start, pIndex - 1);
        quickSort(arr, pIndex + 1, end);
    }
}

function partition(arr, start, end) {
    let pIndex = start;
    const pivot = arr[end];

    for (let i = start; i < end; i++) {
        if (arr[i] <= pivot) {
            let temp = arr[pIndex];
            arr[pIndex] = arr[i];
            arr[i] = temp;

            pIndex += 1;
        }
    }

    let temp = arr[end];
    arr[end] = arr[pIndex];
    arr[pIndex] = temp;

    return pIndex;
}

if (require.main === module) {
    let arr = [8, 9, 7, 5, 2, 4, 6, 1, 3];
    quickSort(arr, 0, arr.length - 1);
    console.log(arr);
}

File: sort/quick/quicksort.cpp
#include <vector>
#include <math.h>

using namespace std;

void quickSort(vector<int>*, int, int);
int partition(vector<int>*, int, int);

void quickSort(vector<int>* arr, int start, int end) {
    if (start < end) {
        int pIdx = partition(arr, start, end);
        quickSort(arr, start, pIdx - 1);
        quickSort(arr, pIdx + 1, end);
    }

}
int partition(vector<int>* arr, int start, int end) {
    vector<int>& A = *arr;
    int pIdx = start;
    int pivot = A[end];

    for (int i = start; i < end; i++) {
        if (A[i] <= pivot) {
            swap(A[pIdx], A[i]);
            pIdx++;
        }
    }
    swap(A[pIdx], A[end]);
    return pIdx;
}

//vector<int> nums { 8, 9, 7, 5, 2, 4, 6, 1, 3, -1 };
//quickSort(&nums, 0, (int)nums.size() - 1);
//printVector(nums);
